File: app/src/main/java/com/sfyc/simple/SimpleActivity.java
package com.sfyc.simple;

import android.content.DialogInterface;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.Toast;

import com.kyleduo.switchbutton.SwitchButton;
import com.larswerkman.lobsterpicker.OnColorListener;
import com.larswerkman.lobsterpicker.sliders.LobsterShadeSlider;
import com.sfyc.ctpv.CountTimeProgressView;

import org.adw.library.widgets.discreteseekbar.DiscreteSeekBar;


/**
 * Author :leilei on 2016/12/20 0115.
 */
public class SimpleActivity extends AppCompatActivity {

    private CountTimeProgressView countTimeProgressView;
    private int checkedItem = 0;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_simple);
        countTimeProgressView = (CountTimeProgressView) findViewById(R.id.countTimeProgressView);
        findViewById(R.id.tv_countTime_style).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                new AlertDialog.Builder(SimpleActivity.this).setTitle("CountTime Style").setSingleChoiceItems(
                        new String[]{"JUMP", "SECOND", "CLOCK", "NONE"}, checkedItem,
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                checkedItem = which;
                                switch (which) {
                                    case 0:
                                        countTimeProgressView.setTextStyle(CountTimeProgressView.TextStyle.JUMP);
                                        countTimeProgressView.startCountTimeAnimation();
                                        dialog.dismiss();
                                        break;
                                    case 1:
                                        countTimeProgressView.setTextStyle(CountTimeProgressView.TextStyle.SECOND);
                                        countTimeProgressView.startCountTimeAnimation();
                                        dialog.dismiss();
                                        break;
                                    case 2:
                                        countTimeProgressView.setTextStyle(CountTimeProgressView.TextStyle.CLOCK);
                                        countTimeProgressView.startCountTimeAnimation();
                                        dialog.dismiss();
                                        break;
                                    case 3:
                                        countTimeProgressView.setTextStyle(CountTimeProgressView.TextStyle.NONE);
                                        countTimeProgressView.startCountTimeAnimation();
                                        dialog.dismiss();
                                        break;
                                    default:
                                        dialog.dismiss();
                                        break;
                                }
                            }
                        }).show();
            }
        });
        // start_angle
        ((DiscreteSeekBar) findViewById(R.id.seek_bar_start_angle)).setOnProgressChangeListener(new DiscreteSeekBar.OnProgressChangeListener() {
            @Override
            public void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser) {

            }

            @Override
            public void onStartTrackingTouch(DiscreteSeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(DiscreteSeekBar seekBar) {
                countTimeProgressView.setStartAngle(seekBar.getProgress());
                countTimeProgressView.startCountTimeAnimation();
            }
        });

        // seek_bar_countTime
        ((DiscreteSeekBar) findViewById(R.id.seek_bar_countTime)).setOnProgressChangeListener(new DiscreteSeekBar.OnProgressChangeListener() {
            @Override
            public void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser) {
//                countTimeProgressView.setCountTime(value * 1000);
//                countTimeProgressView.startCountTimeAnimation();
            }

            @Override
            public void onStartTrackingTouch(DiscreteSeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(DiscreteSeekBar seekBar) {
                countTimeProgressView.setCountTime(seekBar.getProgress() * 1000);
                countTimeProgressView.startCountTimeAnimation();
            }
        });
        // Border
        ((DiscreteSeekBar) findViewById(R.id.seekbar_border_width)).setOnProgressChangeListener(new DiscreteSeekBar.OnProgressChangeListener() {
            @Override
            public void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser) {

            }

            @Override
            public void onStartTrackingTouch(DiscreteSeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(DiscreteSeekBar seekBar) {

                countTimeProgressView.setBorderWidth(seekBar.getProgress());
            }
        });

        // setBorderBottomColor
        ((LobsterShadeSlider) findViewById(R.id.shadeslider_bottom_color)).addOnColorListener(new OnColorListener() {
            @Override
            public void onColorChanged(@ColorInt int color) {
                countTimeProgressView.setBorderBottomColor(color);
            }

            @Override
            public void onColorSelected(@ColorInt int color) {
            }
        });

        //setBorderDrawColor
        ((LobsterShadeSlider) findViewById(R.id.shadeslider_draw_color)).addOnColorListener(new OnColorListener() {
            @Override
            public void onColorChanged(@ColorInt int color) {
                countTimeProgressView.setBorderDrawColor(color);
            }

            @Override
            public void onColorSelected(@ColorInt int color) {
            }
        });
        //setBorderDrawColor
        ((LobsterShadeSlider) findViewById(R.id.shadeslider_bg_color)).addOnColorListener(new OnColorListener() {
            @Override
            public void onColorChanged(@ColorInt int color) {
                countTimeProgressView.setBackgroundColor(color);
            }

            @Override
            public void onColorSelected(@ColorInt int color) {
            }
        });

        ((DiscreteSeekBar) findViewById(R.id.seekbar_ball_width)).setOnProgressChangeListener(new DiscreteSeekBar.OnProgressChangeListener() {
            @Override
            public void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser) {

            }

            @Override
            public void onStartTrackingTouch(DiscreteSeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(DiscreteSeekBar seekBar) {
                countTimeProgressView.setMarkBallWidth(seekBar.getProgress());
            }
        });
        // setBorderBottomColor
        ((LobsterShadeSlider) findViewById(R.id.shadeslider_ball_color)).addOnColorListener(new OnColorListener() {
            @Override
            public void onColorChanged(@ColorInt int color) {
                countTimeProgressView.setMarkBallColor(color);
            }

            @Override
            public void onColorSelected(@ColorInt int color) {
            }
        });

        SwitchButton sb = (SwitchButton) findViewById(R.id.sb_ball_flag);
        sb.setChecked(true);
        sb.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                countTimeProgressView.setMarkBallFlag(isChecked);
            }
        });

        countTimeProgressView.setStartAngle(0);
        countTimeProgressView.setCountTime(6000);
        countTimeProgressView.setTextStyle(CountTimeProgressView.TextStyle.SECOND);
        countTimeProgressView.setBorderWidth(8);
        countTimeProgressView.setBorderBottomColor(Color.GRAY);
        countTimeProgressView.setBorderDrawColor(Color.RED);
        countTimeProgressView.setBackgroundColor(Color.WHITE);
        countTimeProgressView.setMarkBallFlag(true);
        countTimeProgressView.setMarkBallWidth(12);
        countTimeProgressView.setMarkBallColor(Color.GREEN);

        countTimeProgressView.addOnEndListener(new CountTimeProgressView.OnEndListener() {
            @Override
            public void onAnimationEnd() {
                Toast.makeText(SimpleActivity.this, "时间到", Toast.LENGTH_SHORT).show();
            }
            @Override
            public void onClick(long overageTime) {
                if (countTimeProgressView.isRunning()) {
                    countTimeProgressView.cancelCountTimeAnimation();
                    Log.e("overageTime","overageTime = "+overageTime);
                } else {
                    countTimeProgressView.startCountTimeAnimation();
                }
            }
        });
        countTimeProgressView.startCountTimeAnimation();
    }
}

File: app/src/main/java/com/sfyc/simple/SimpleActivity.kt
package com.sfyc.simple

import android.graphics.Color
import android.graphics.Paint
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.SeekBar
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.switchmaterial.SwitchMaterial
import com.sfyc.ctpv.CountTimeProgressView

class SimpleActivity : AppCompatActivity() {

    private lateinit var countTimeProgressView: CountTimeProgressView
    private var checkedItem = 0
    private var strokeCapIndex = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_simple)
        countTimeProgressView = findViewById(R.id.countTimeProgressView)

        setupTextStyleSelector()
        setupSeekBars()
        setupSwitches()
        setupStrokeCapSelector()
        initCountTimeView()
    }

    private fun setupTextStyleSelector() {
        findViewById<View>(R.id.tv_countTime_style).setOnClickListener {
            AlertDialog.Builder(this)
                .setTitle(R.string.dialog_count_time_style)
                .setSingleChoiceItems(arrayOf("JUMP", "SECOND", "CLOCK", "NONE"), checkedItem) { dialog, which ->
                    checkedItem = which
                    when (which) {
                        0 -> {
                            countTimeProgressView.textStyle = CountTimeProgressView.TEXT_STYLE_JUMP
                            countTimeProgressView.titleCenterText = getString(R.string.text_skip)
                        }
                        1 -> {
                            countTimeProgressView.textStyle = CountTimeProgressView.TEXT_STYLE_SECOND
                            countTimeProgressView.titleCenterText =
                                getString(R.string.countdown_skip_seconds_template)
                        }
                        2 -> countTimeProgressView.textStyle = CountTimeProgressView.TEXT_STYLE_CLOCK
                        3 -> countTimeProgressView.textStyle = CountTimeProgressView.TEXT_STYLE_NONE
                    }
                    countTimeProgressView.startCountTimeAnimation()
                    dialog.dismiss()
                }
                .show()
        }
    }

    private fun setupStrokeCapSelector() {
        val capNames = arrayOf("BUTT", "ROUND", "SQUARE")
        val caps = arrayOf(Paint.Cap.BUTT, Paint.Cap.ROUND, Paint.Cap.SQUARE)
        val tvStrokeCap = findViewById<View>(R.id.tv_stroke_cap)

        tvStrokeCap.setOnClickListener {
            AlertDialog.Builder(this)
                .setTitle(R.string.dialog_stroke_cap)
                .setSingleChoiceItems(capNames, strokeCapIndex) { dialog, which ->
                    strokeCapIndex = which
                    countTimeProgressView.strokeCap = caps[which]
                    countTimeProgressView.startCountTimeAnimation()
                    dialog.dismiss()
                }
                .show()
        }
    }

    private fun setupSeekBars() {
        findViewById<SeekBar>(R.id.seek_bar_countTime).setOnSeekBarChangeListener(
            onStopListener { progress ->
                countTimeProgressView.countTime = progress.coerceAtLeast(1) * 1000L
                countTimeProgressView.startCountTimeAnimation()
            }
        )

        findViewById<SeekBar>(R.id.seek_bar_start_angle).setOnSeekBarChangeListener(
            onStopListener { progress ->
                countTimeProgressView.startAngle = progress.toFloat()
                countTimeProgressView.startCountTimeAnimation()
            }
        )

        findViewById<SeekBar>(R.id.seekbar_ball_width).setOnSeekBarChangeListener(
            onStopListener { progress ->
                countTimeProgressView.markBallWidth = progress.toFloat()
            }
        )

        findViewById<SeekBar>(R.id.seekbar_border_width).setOnSeekBarChangeListener(
            onStopListener { progress ->
                countTimeProgressView.borderWidth = progress.toFloat()
            }
        )
    }

    private fun setupSwitches() {
        findViewById<SwitchMaterial>(R.id.sb_ball_flag).setOnCheckedChangeListener { _, isChecked ->
            countTimeProgressView.markBallFlag = isChecked
        }

        findViewById<SwitchMaterial>(R.id.sb_clockwise).setOnCheckedChangeListener { _, isChecked ->
            countTimeProgressView.clockwise = isChecked
            countTimeProgressView.startCountTimeAnimation()
        }

        findViewById<SwitchMaterial>(R.id.sb_gradient).setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                countTimeProgressView.setGradientColors(
                    Color.parseColor("#FF6B6B"),
                    Color.parseColor("#4ECDC4")
                )
            } else {
                countTimeProgressView.gradientStartColor = 0
                countTimeProgressView.gradientEndColor = 0
                countTimeProgressView.borderDrawColor = Color.RED
            }
            countTimeProgressView.startCountTimeAnimation()
        }
    }

    private fun initCountTimeView() {
        with(countTimeProgressView) {
            startAngle = 0f
            countTime = 6000L
            textStyle = CountTimeProgressView.TEXT_STYLE_SECOND
            borderWidth = 4f
            borderBottomColor = Color.GRAY
            borderDrawColor = Color.RED
            backgroundColorCenter = Color.WHITE
            markBallFlag = true
            markBallWidth = 6f
            markBallColor = Color.GREEN
            titleCenterText = getString(R.string.countdown_skip_seconds_template)
            titleCenterTextColor = Color.BLACK
            titleCenterTextSize = 16f

            warningTime = 3000L
            warningColor = Color.parseColor("#FF3B30")
            clickableAfterMillis = 2000L
            disabledText = getString(R.string.text_please_wait)

            setOnStateChangedListener { state ->
                Log.d(TAG, "State changed: $state")
            }

            setOnTickListener { remainingMillis, remainingSeconds ->
                Log.d(TAG, "Tick: ${remainingSeconds}s ($remainingMillis ms)")
            }

            setOnWarningListener { remainingMillis ->
                Log.d(TAG, "Warning threshold reached, remaining=${remainingMillis}ms")
            }

            setOnCountdownEndListener {
                Toast.makeText(this@SimpleActivity, getString(R.string.toast_times_up), Toast.LENGTH_SHORT).show()
            }

            setOnClickCallback { overageTime ->
                if (isRunning) {
                    cancelCountTimeAnimation()
                    Log.d(TAG, "Click canceled, remaining=$overageTime ms")
                } else {
                    startCountTimeAnimation()
                }
            }

            addOnProgressChangedListener { progress, remainingMillis ->
                Log.d(TAG, "progress=$progress, remaining=$remainingMillis")
            }

            bindLifecycle(this@SimpleActivity)
            startCountTimeAnimation()
        }
    }

    private fun onStopListener(action: (Int) -> Unit): SeekBar.OnSeekBarChangeListener {
        return object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) = Unit
            override fun onStartTrackingTouch(seekBar: SeekBar) = Unit
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                action(seekBar.progress)
            }
        }
    }

    companion object {
        private const val TAG = "SimpleActivity"
    }
}
